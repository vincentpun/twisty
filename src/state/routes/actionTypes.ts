export enum RoutesActionsTypes {
  CreateQuery = 'ROUTES_QUERY_CREATE',
  FetchRoutesToken = 'ROUTES_QUERY_FETCH_ROUTES',
  FetchRoutesTokenReceive = 'ROUTES_QUERY_FETCH_ROUTES_TOKEN_RECEIVE',
  PollResult = 'ROUTES_QUERY_POLL_RESULT',
  PollResultReceiveInProgress = 'ROUTES_QUERY_POLL_RESULT_RECEIVE_IN_PROGRESS',
  QuerySuccess = 'ROUTES_QUERY_SUCCESS',
  QueryReceive = 'ROUTES_QUERY_RECEIVE',
  QueryError = 'ROUTES_QUERY_ERROR',
}
