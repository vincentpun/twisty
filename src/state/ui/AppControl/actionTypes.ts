export enum AppControlUIActionTypes {
  Select = 'APP_CONTROL_SELECT',

  FetchClientLocationRequest = 'APP_CONTROL_FETCH_CLIENT_LOCATION_REQUEST',
  FetchClientLocationReceive = 'APP_CONTROL_FETCH_CLIENT_LOCATION_RECEIVE',
  FetchClientLocationRequestSuccess = 'APP_CONTROL_FETCH_CLIENT_LOCATION_REQUEST_SUCCESS',
  FetchClientLocationRequestError = 'APP_CONTROL_FETCH_CLIENT_LOCATION_REQUEST_ERROR',

  PickStartingPoint = 'APP_CONTROL_PICK_STARTING_POINT',
  RemoveCoordinates = 'APP_CONTROL_REMOVE_COORDINATES',

  AddDropoffLocation = 'APP_CONTROL_ADD_DROPOFF_LOCATION',
}
