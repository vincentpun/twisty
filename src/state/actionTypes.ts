import { UIActionTypes } from './ui/actionTypes';
import { RoutesActionsTypes } from './routes/actionTypes';

export type ActionTypes = UIActionTypes | RoutesActionsTypes;
