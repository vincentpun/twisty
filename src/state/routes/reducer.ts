import { combineReducers } from 'redux';

import { Actions } from './actions';
import { RoutesActionsTypes } from './actionTypes';
import { MapCoordinates } from '../ui/AppControl/types';

export enum RouteQueryStatus {
  NotStarted = 'ROUTE_QUERY_STATUS_NOT_STARTED',
  FetchingRoutesToken = 'ROUTE_QUERY_STATUS_FETCHING_ROUTES_TOKEN',
  ToPoll = 'ROUTE_QUERY_STATUS_TO_POLL',
  PollingInProgress = 'ROUTE_QUERY_STATUS_POLLING_IN_PROGRESS',
  Success = 'ROUTE_QUERY_STATUS_SUCCESS',
  Error = 'ROUTE_QUERY_STATUS_ERROR',
}

export interface RoutesState {
  queries: {
    [id: string]: RouteQuery;
  };
}

export interface RouteQuery {
  id: string;
  status: RouteQueryStatus;
  token?: string;
  response?: {
    waypoints: MapCoordinates[];
    totalDistance: number;
    totalTime: number;
    [key: string]: any;
  }
  error?: any;
}

const initialState = {
  queries: {} as { [id: string]: RouteQuery },
};

export const queries = (state = initialState.queries, action: Actions) => {
  switch (action.type) {
    case RoutesActionsTypes.CreateQuery: {
      const id = action.payload;
      return {
        ...state,
        [id]: {
          id,
          status: RouteQueryStatus.NotStarted,
        },
      };
    }
    case RoutesActionsTypes.FetchRoutesToken: {
      const id = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.FetchingRoutesToken,
        },
      };
    }
    case RoutesActionsTypes.FetchRoutesTokenReceive: {
      const { id, token } = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.ToPoll,
          token,
        },
      };
    };
    case RoutesActionsTypes.PollResult: {
      const id = action.payload;

      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.PollingInProgress,
        },
      };
    }
    case RoutesActionsTypes.PollResultReceiveInProgress: {
      const id = action.payload;

      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.ToPoll,
        },
      };
    }
    case RoutesActionsTypes.QuerySuccess: {
      const id = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.Success,
        },
      };
    }
    case RoutesActionsTypes.QueryReceive: {
      const { id, response } = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          response,
        }
      };
    }
    case RoutesActionsTypes.QueryError: {
      const { id, error } = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.Error,
          error,
        }
      };
    }
    default:
      return state;
  }
}

export default combineReducers<RoutesState>({
  queries,
});
