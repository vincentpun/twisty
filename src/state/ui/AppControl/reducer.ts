import { combineReducers } from 'redux';

import { createAction } from 'src/state/actions';
import { AppControlUIActionTypes } from './actionTypes';
import { AppControlSelection } from './types';
import { Actions } from './actions';

export interface AppControlUIState {
  selection: AppControlSelection;
}

const initialState: AppControlUIState = {
  selection: null,
};

export const selection = (state = initialState.selection, action: Actions) => {
  switch (action.type) {
    case AppControlUIActionTypes.Select:
      return action.payload;
    default:
      return state;
  }
}

export default combineReducers<AppControlUIState>({
  selection,
});
