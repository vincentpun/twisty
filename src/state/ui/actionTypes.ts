import { AppControlUIActionTypes } from './AppControl/actionTypes';

export enum AppUIActionTypes {
  SwitchMode = 'APP_UI_SWITCH_MODE',
}

export type UIActionTypes = AppUIActionTypes | AppControlUIActionTypes;
