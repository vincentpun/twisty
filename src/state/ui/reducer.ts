import { combineReducers } from 'redux';

import appControl, { AppControlUIState } from './AppControl/reducer';
import routes, { RoutesState } from './Routes/reducer';
import { Actions } from './actions';
import { AppUIActionTypes } from './actionTypes';

export interface UIState {
  mode: Readonly<AppUIMode>,
  routes: Readonly<RoutesState>,
  appControl: Readonly<AppControlUIState>,
}

export enum AppUIMode {
  AppControl = 'appControl',
  Route = 'route',
}

export const mode = (state = AppUIMode.AppControl, action: Actions) => {
  switch (action.type) {
    case AppUIActionTypes.SwitchMode:
      return action.payload;
    default:
      return state;
  }
}

export default combineReducers<UIState>({
  appControl,
  routes,
  mode,
});
