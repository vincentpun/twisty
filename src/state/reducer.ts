import { combineReducers } from 'redux';

import ui, { UIState } from './ui/reducer';
import routes, { RoutesState } from './routes/reducer';

export interface State {
  ui: Readonly<UIState>
  routes: Readonly<RoutesState>,
}

export default combineReducers<State>({
  ui,
  routes,
});
