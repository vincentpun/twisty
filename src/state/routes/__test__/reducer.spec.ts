import * as reducer from '../reducer';
import { RoutesActionsTypes } from '../actionTypes';
import { RouteQueryStatus } from '../reducer';

describe('routes reducer', () => {
  it('matches the snapshot', () => {
    const state = reducer.default(undefined, { type: null });
    expect(state).toMatchSnapshot();
  });

  describe('queries', () => {
    it('matches the snapshot', () => {
      const state = reducer.queries(undefined, { type: null, payload: null });
      expect(state).toMatchSnapshot();
    });

    it('handles RoutesActionsTypes.CreateQuery', () => {
      let state = reducer.queries({}, {
        type: RoutesActionsTypes.CreateQuery,
        payload: 'id',
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.NotStarted,
        },
      });

      state = reducer.queries({
        id2: {
          id: 'id2',
          status: RouteQueryStatus.NotStarted,
        },
      }, {
        type: RoutesActionsTypes.CreateQuery,
        payload: 'id',
      });
      expect(state).toEqual({
        id2: {
          id: 'id2',
          status: RouteQueryStatus.NotStarted,
        },
        id: {
          id: 'id',
          status: RouteQueryStatus.NotStarted,
        },
      });
    });

    it('handles RoutesActionsTypes.FetchRoutesToken', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.NotStarted,
        }
      }, {
        type: RoutesActionsTypes.FetchRoutesToken,
        payload: 'id',
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.FetchingRoutesToken,
        },
      });
    });

    it('handles RoutesActionsTypes.FetchRoutesTokenReceive', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.FetchingRoutesToken,
        }
      }, {
        type: RoutesActionsTypes.FetchRoutesTokenReceive,
        payload: {
          id: 'id',
          token: 'test',
        },
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.ToPoll,
          token: 'test',
        },
      });
    });

    it('handles RoutesActionsTypes.PollResult', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.ToPoll,
          token: 'test',
        }
      }, {
        type: RoutesActionsTypes.PollResult,
        payload: 'id',
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.PollingInProgress,
          token: 'test',
        },
      });
    });

    it('handles RoutesActionsTypes.PollResultReceiveInProgress', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.PollingInProgress,
          token: 'test',
        }
      }, {
        type: RoutesActionsTypes.PollResultReceiveInProgress,
        payload: 'id',
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.ToPoll,
          token: 'test',
        },
      });
    });

    it('handles RoutesActionsTypes.QuerySuccess', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.ToPoll,
          token: 'test',
        }
      }, {
        type: RoutesActionsTypes.QuerySuccess,
        payload: 'id',
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.Success,
          token: 'test',
        },
      });
    });

    it('handles RoutesActionsTypes.QueryReceive', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.Success,
          token: 'test',
        }
      }, {
        type: RoutesActionsTypes.QueryReceive,
        payload: {
          id: 'id',
          response: {
            waypoints: [],
            totalDistance: 10,
            totalTime: 10,
          },
        },
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.Success,
          token: 'test',
          response: {
            waypoints: [],
            totalDistance: 10,
            totalTime: 10,
          },
        },
      });
    });

    it('handles RoutesActionsTypes.QueryError', () => {
      const state = reducer.queries({
        id: {
          id: 'id',
          status: RouteQueryStatus.ToPoll,
          token: 'test',
        }
      }, {
        type: RoutesActionsTypes.QueryError,
        payload: {
          id: 'id',
          error: 'test',
        },
      });
      expect(state).toEqual({
        id: {
          id: 'id',
          status: RouteQueryStatus.Error,
          token: 'test',
          error: 'test',
        },
      });
    });
  });
});
