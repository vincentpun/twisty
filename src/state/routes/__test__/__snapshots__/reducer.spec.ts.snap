// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes reducer matches the snapshot 1`] = `
Object {
  "queries": Object {},
}
`;

exports[`routes reducer queries matches the snapshot 1`] = `Object {}`;
